FEATURE: Identity

osmanaliyardim opened the pull request:
### [Task 1](https://learn.epam.com/myLearning/program?groupGuid=68c6d8bc-9684-4ee5-a549-048b3ef481d8) - Local user database and ASP.Net Core Identity

### What this change does
> Focus on why the change was made and provide short, clean and good summary
* Add idendity controls.

### Code changes
> Explain what files and code parts were changed in scope of PR and reviewer should pay attention for
* Added registration, login and password reset functionality.

### What has been done to confirm PR works and does not break existing stuff
> Provide guarantees that your change does not ruin existing functionality and works as expected
* Tested on localhost.

### Screenshots
> Screenshots with descriptions confirming your correct results
* Password reset mail (Forgot password)
![PasswordReset](https://user-images.githubusercontent.com/25593515/204493041-75191006-0b12-4863-9bf4-7e2b8f45c323.png)

osmanaliyardim committed:
FEATURE: Identity

Login, logout, registration and password reset functionalities scaffolded and applied.

dotnet-fizzyy commented on WebNetSample/appsettings.json:
```diff
@@ -9,5 +9,14 @@
+    "port": 465, //587
+    "emailName": "TestNameEmail",
+    "emailAddress": "osmantest60@gmail.com",
+    "password": "aencglekznlcgqur",
```
I really want to believe this password is not a real one :)
Whenever you need to leave some sensitive data in code (like app settings), make sure you didn't commit them and they do not exist in git history. For app settings files you may leave just some examples of password, like `Your_password`

osmanaliyardim replied:
Well, it was a fake password but still removed :)

dotnet-fizzyy commented on WebNetSample/appsettings.json:
```diff
@@ -9,5 +9,14 @@
   },
+  "smtpClientSettings": {
+    "host": "smtp.gmail.com",
+    "port": 465, //587
```
Let's remove commented item here

osmanaliyardim committed:
FEATURE: Azure AD

Login and registering option with Azure AD has applied.

dotnet-fizzyy commented on Business/BusinessServiceRegistration.cs:
```diff
@@ -0,0 +1,15 @@
+    {
+        services.AddSingleton(appSettings);
+
+        return services;
```
I think no need to return here value since `IServiceCollection` is a reference type

dotnet-fizzyy commented on Business/Startup/SmtpClientSettings.cs:
```diff
@@ -0,0 +1,24 @@
+
+    public bool UseSsl { get; set; }
+
+    public void Validate()
```
Good idea to add validation of your app settings!
However, I am not really sure it works since you didn't use any data annotation attributes above your properties

dotnet-fizzyy commented on DataAccess/Concrete/EntityFramework/Contexts/WebNetSampleDBContext.cs:
```diff
@@ -25,5 +27,7 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
         modelBuilder.ApplyConfiguration(new CategoryConfiguration());
         modelBuilder.ApplyConfiguration(new ProductConfiguration());
         modelBuilder.ApplyConfiguration(new SupplierConfiguration());
```
No need to use these members since you are reading all configs [here](https://github.com/osmanaliyardim/WebNetSample/blob/6eaa349a52395d115f7070598b5775cd55727fc3/DataAccess/Concrete/EntityFramework/Contexts/WebNetSampleDBContext.cs#L25).

dotnet-fizzyy commented on DataAccess/Concrete/EntityFramework/Contexts/WebNetSampleDBContext.cs:
```diff
@@ -25,5 +27,7 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
         modelBuilder.ApplyConfiguration(new ProductConfiguration());
         modelBuilder.ApplyConfiguration(new SupplierConfiguration());
+
+        base.OnModelCreating(modelBuilder);
```
I strongly advise to put call of base method at the beginning since it may override your custom configs (based on my own experience, had to spent about an hour to determine why it was not working)

dotnet-fizzyy commented on WebNetSample/Areas/Identity/Pages/Account/ConfirmEmail.cshtml.cs:
```diff
@@ -0,0 +1,51 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+#nullable disable
```
No need to store autogenerated comments since, let's remove it from the rest of files, please

dotnet-fizzyy commented on WebNetSample/Areas/Identity/Pages/Account/ConfirmEmail.cshtml.cs:
```diff
@@ -0,0 +1,51 @@
+
+        /// <summary>
+        ///     This API supports the ASP.NET Core Identity default UI infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
```
No need to have this extra indentation. Also, I think, we can remove such comments since they don't bring any useful information

dotnet-fizzyy commented on WebNetSample/Areas/Identity/Pages/Account/ConfirmEmail.cshtml.cs:
```diff
@@ -0,0 +1,51 @@
+        public string StatusMessage { get; set; }
+        public async Task<IActionResult> OnGetAsync(string userId, string code)
+        {
+            if (userId == null || code == null)
```
Know, this code is autogenerated, it should be refactored as well to make your codebase better and consistent. Let's use `string.IsNullOrWhiteSpace`

dotnet-fizzyy commented on WebNetSample/Areas/Identity/Pages/Account/ConfirmEmail.cshtml.cs:
```diff
@@ -0,0 +1,51 @@
+            code = Encoding.UTF8.GetString(WebEncoders.Base64UrlDecode(code));
+            var result = await _userManager.ConfirmEmailAsync(user, code);
+            StatusMessage = result.Succeeded ? "Thank you for confirming your email." : "Error confirming your email.";
+            return Page();
```
Don't forget about empty line before return statement

dotnet-fizzyy commented on WebNetSample/Areas/Identity/Pages/Account/ExternalLogin.cshtml.cs:
```diff
@@ -0,0 +1,223 @@
+        /// <summary>
+        ///     This API supports the ASP.NET Core Identity default UI infrastructure and is not intended to be used
+        ///     directly from your code. This API may change or be removed in future releases.
+        /// </summary>
```
We can remove it here since it does not contain any useful information

dotnet-fizzyy commented on WebNetSample/Areas/Identity/Pages/Account/ExternalLogin.cshtml.cs:
```diff
@@ -0,0 +1,223 @@
+            }
+
+            // Sign in the user with this external login provider if the user already has a login.
+            var result = await _signInManager.ExternalLoginSignInAsync(info.LoginProvider, info.ProviderKey, isPersistent: false, bypassTwoFactor: true);
```
Let's align args into a column

dotnet-fizzyy commented on WebNetSample/Areas/Identity/Pages/Account/ExternalLogin.cshtml.cs:
```diff
@@ -0,0 +1,223 @@
+                {
+                    result = await _userManager.AddLoginAsync(user, info);
+                    if (result.Succeeded)
+                    {
```
Let's refactor these huge `if` nesting, it makes code really difficult to read

dotnet-fizzyy commented on WebNetSample/Areas/Identity/Pages/Account/Register.cshtml.cs:
```diff
@@ -0,0 +1,180 @@
+                var result = await _userManager.CreateAsync(user, Input.Password);
+
+                if (result.Succeeded)
+                {
```
Same stuff with huge nesting

dotnet-fizzyy reviewed (commented):
There are a lot of autogenerated files.
We will not spent too much time for these code parts, let's just remove extra comments, big code nesting and change some variable names to make it a little bit better. The rest of part will be refactored in scope of `Issues` part

osmanaliyardim committed:
FEATURE: Authorization

Users controller has been created and authorization has applied to acces it.

osmanaliyardim committed:
REFACTOR: Namings

Role namings have been made constants.

osmanaliyardim committed:
REFACTOR: Tabulation

Tabulation has been applied to increase readability.

osmanaliyardim committed:
Delete appsettings.json

Because of sensitive info

osmanaliyardim committed:
REFACTOR: Readability

Nestings, namings, comments and tabulations have been rearrenged to increase readability.

osmanaliyardim committed:
REFACTOR: Remove return

Unneccessary return statement removed because of reference type.

dotnet-fizzyy commented on WebNetSample/Areas/Identity/Pages/Account/Login.cshtml.cs:
```diff
@@ -1,140 +1,107 @@
         /// <summary>
         ///     This API supports the ASP.NET Core Identity default UI infrastructure and is not intended to be used
         ///     directly from your code. This API may change or be removed in future releases.
         /// </summary>
```
Let's remove it as well, please (for the other members as well)

dotnet-fizzyy commented on WebNetSample/Areas/Identity/Pages/Account/Manage/DownloadPersonalData.cshtml.cs:
```diff
@@ -1,67 +1,56 @@
+        var personalData = new Dictionary<string, string>();
+        var personalDataProps = typeof(IdentityUser).GetProperties().Where(
+                        prop => Attribute.IsDefined(prop, typeof(PersonalDataAttribute)));
+        foreach (var p in personalDataProps)
```
Let's give more clear name instead of `p` and `l` below, please

osmanaliyardim committed:
REFACTOR: Removing Comments

osmanaliyardim committed:
Merge pull request #22 from osmanaliyardim/feature/security-task-3

FEATURE: Authorization

osmanaliyardim committed:
Merge pull request #21 from osmanaliyardim/feature/security-task-2

FEATURE: Azure AD
